### PR TITLE
chore(Profiling): add infrastructure for supporting timeline, but timeline isn't supported yet

### PR DESCRIPTION
**Pull request by sanchda**
Next week I'll finally add support for testing some of these new interface.  Moreover, at the current time this stuff is pretty much only for internal testing, so I'll document when we're closer to going public.

The, uh, framing for this is kind of weird.  I'm sorry about that.

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/internal/datadog/profiling/ddup/_ddup.pyx [hunk 1]
```diff
@@ -234,7 +235,6 @@ cdef class SampleHandle:
     def push_task_id(self, task_id: Optional[int]) -> None:
         if self.ptr is not NULL:
             if task_id is not None:
-                ddup_push_task_id(self.ptr, task_id)
```
Comment (taegyunkim) [inline ddtrace/internal/datadog/profiling/ddup/_ddup.pyx:237]: Why is this line deleted?

Comment (sanchda) [reply ddtrace/internal/datadog/profiling/ddup/_ddup.pyx:237]: Sorry, it's a weird artifact of the merge.  I actually replaced it with the line below (line 242)
```
ddup_push_task_id(self.ptr, clamp_to_int64_unsigned(task_id))
```

**Review by emmettbutler — APPROVED** [review]

File: ddtrace/internal/datadog/profiling/dd_wrapper/src/sample.cpp [hunk 2]
```diff
@@ -316,6 +318,35 @@ Datadog::Sample::push_class_name(std::string_view class_name)
...
+    // Monotonic times have their epoch at the system start, so they need an
+    // adjustment to the standard epoch
+    // Just set a static for now and use a lambda to compute the offset once
+    const static auto offset = []() {
+        // Get the current epoch time
+        using namespace std::chrono;
+        auto epoch_ns = duration_cast<nanoseconds>(system_clock::now().time_since_epoch()).count();
+
+        // Get the current monotonic time.  Use clock_gettime directly because the standard underspecifies
+        // which clock is actually used in std::chrono
```
Comment (taegyunkim) [inline ddtrace/internal/datadog/profiling/dd_wrapper/src/sample.cpp:333]: Looks like `std::chrono::steady_clock` seems to be using `clock_gettime(CLOCK_MONOTONIC, ..)` under the hood. Can this be used instead? or was there a reason to use clock_gettime instead of std::chrono one?

File: ddtrace/internal/datadog/profiling/dd_wrapper/src/sample.cpp [hunk 3]
```diff
@@ -316,6 +318,35 @@ Datadog::Sample::push_class_name(std::string_view class_name)
     return true;
 }
 
+bool
+Datadog::Sample::push_monotonic_ns(int64_t _monotonic_ns)
```
Comment (taegyunkim) [inline ddtrace/internal/datadog/profiling/dd_wrapper/src/sample.cpp:322]: Could someone provide me some context why functions here are named `push_*` instead of `add_*` or something else?

**Review by taegyunkim — APPROVED** [review]
Generally LGTM.

File: ddtrace/internal/datadog/profiling/dd_wrapper/src/sample.cpp (hunk 3 above)
Comment (sanchda) [reply ddtrace/internal/datadog/profiling/dd_wrapper/src/sample.cpp:322]: There's no _great_ reason, but I chose "push" instead of "add" because

1. Data transfers into libdatadog is mediated through an intermediate Sample object; we haven't committed the data until that Sample gets flushed.  As much as we "add" to the Sample, I was hoping that, on the Python side, there's this understanding that data isn't really saved until the Sample has been flushed.
2. State crosses an FFI boundary--we seldom, if ever, manipulate storage directly.

Happy to talk about this.  It's not set in stone.

File: ddtrace/internal/datadog/profiling/dd_wrapper/src/interface.cpp [hunk 4]
```diff
@@ -99,12 +99,19 @@ ddup_config_sample_type(unsigned int _type) // cppcheck-suppress unusedFunction
...
 }
+
 void
 ddup_config_max_nframes(int max_nframes) // cppcheck-suppress unusedFunction
 {
     Datadog::SampleManager::set_max_nframes(max_nframes);
 }
 
+void
+ddup_config_timeline_enabled(bool enabled) // cppcheck-suppress unusedFunction
```
Comment (danielsn) [inline ddtrace/internal/datadog/profiling/dd_wrapper/src/interface.cpp:110]: Grammatically, I'd call this `enable_timeline` rather than the current, which to me implies a predicate?

**Review by danielsn — APPROVED** [review]
Couple questions.  LGTM

File: ddtrace/internal/datadog/profiling/dd_wrapper/include/sample.hpp [hunk 5]
```diff
@@ -36,6 +36,10 @@ class Sample
     // Storage for values
     std::vector<int64_t> values = {};
 
+    // Additional metadata
+    bool timeline_enabled = false;
```
[the remark this answers is not in the document]
Comment (sanchda) [reply ddtrace/internal/datadog/profiling/dd_wrapper/include/sample.hpp]: I elevated this to a static variable and changed the naming slightly.  It's a little awkward right now for short-lived sample objects to refer to outside state, but hopefully we can fix this at some point.
